### DISCUSSION TITLE
Rescaling the shape/body at runtime using WASM/javascript

### DISCUSSION THREAD

**Discussion by MAG-AdrianMeredith** [jrouwe/JoltPhysics#1154] [Q&A] [answered]
Hello all,

I've read the docs on recommendations of how to scale a shape in JOLT however the advice doesn't seem possible to follow when in WASM due to the lack of the ability to access to the ScaledShape from the body.  As referenced by this line 

"		const ScaledShape *scaled_shape = static_cast<const ScaledShape *>(body.GetShape());
		const Shape *non_scaled_shape = scaled_shape->GetInnerShape();"

Javascript does have the capability to cast the shape to a scaled shape.  Is there another way?  I had tried ScaleShape() but that isn't useful as it is cumulative and so the shape gets larger each time.

Many thanks in advance

**Comment by MAG-AdrianMeredith** [comment]
So I've managed to get it to work now with this.  By instead retaining the reference to the original shape and passing that to the scale function (so I don't need to get it from the "ScaledShape->GetInnerShape()"). 

```
const scale = new Jolt.Vec3(...v);
const result = new Jolt.ScaledShape(shape, scale);
bodyInterface.SetShape(body.GetID(), result, true, Jolt.EActivation_Activate);
Jolt.destroy(scale);
```

But I'm not sure about destroying the ScaledShape?  it crashes if I try and remove the previous one?  whats the option here so that I'm not leaking memory?

**Reply by jrouwe** [reply, marked as answer]
> Javascript does have the capability to cast the shape to a scaled shape.

I'm assuming you meant to say 'doesn't have the capability'. Use `Jolt.castObject(shape, Jolt.ScaledShape)` to cast it.

> But I'm not sure about destroying the ScaledShape? it crashes if I try and remove the previous one? whats the option here so that I'm not leaking memory?

As described [here](https://github.com/jrouwe/JoltPhysics.js?tab=readme-ov-file#memory-management), shapes are reference counted. When you assign a shape to a body it gets a ref count of 1. If you then assign another shape to the same body, the first shape is released, its ref count reaches 0 and the shape gets deleted.

**Reply by MAG-AdrianMeredith** [reply]
Amazing thank you for the clarification
